docs/switch.rst

```diff
@@ -0,0 +1,369 @@
+``gdal.SetErrorHandler()`` and ``gdal.UseExceptions()``.
+
+Rasterio has modules that don't require complete initialization and
+configuration of GDAL (``rasterio.dtypes``, ``rasterio.profiles``, and
+``rasterio.windows``, for example) and in the interest of reducing overhead
+doesn't register format drivers and error handlers until they are needed. The
+functions that do need fully initialized GDAL environments will ensure that
+they exist. ``rasterio.open()`` is the foremost of this category of functions.
+Consider the example code below.
+
+.. code-block:: python
+
+   import rasterio
+   # The GDAL environment has no registered format drivers or error
+   # handlers at this point.
+
+   with rasterio.open('example.tif') as src:
+       # Format drivers and error handlers are registered just before
+       # open() executes.
+
+Importing ``rasterio`` does not initialize the GDAL environment. Calling
+``rasterio.open()`` does. This is different from ``gdal`` where ``import
+osgeo.gdal``, not ``osgeo.gdal.Open()``, initializes the GDAL environment.
+
+Rasterio has an abstraction for the GDAL environment, ``rasterio.Env``, that
+can be invoked explicitly for more control over the configuration of GDAL as
+shown below.
+
+.. code-block:: python
+
+   import rasterio
+   # The GDAL environment has no registered format drivers or error
+   # handlers at this point.
+
+   with rasterio.Env(CPL_DEBUG=True, GDAL_CACHEMAX=512):
+       # This ensures that all drivers are registered in the global
+       # context. Within this block *only* GDAL's debugging messages
+       # are turned on and the raster block cache size is set to 512MB.
+   
+       with rasterio.open('example.tif') as src:
+           # Perform GDAL operations in this context.
+           # ...
+           # Done.
+
+   # At this point, configuration options are set back to their
+   # previous (possibly unset) values. The raster block cache size
+   # is returned to its default (5% of available RAM) and debugging
+   # messages are disabled.
+
+As mentioned previously, ``gdal`` has no such abstraction for the GDAL
+environment. The nearest approximation would be something like the code
+below.
+
+.. code-block:: python
+
+   from osgeo import gdal
+
+   # Define a new configuration, save the previous configuration,
+   # and then apply the new one.
+   new_config = {
+       'CPL_DEBUG': 'ON', 'GDAL_CACHEMAX': '512'}
+   prev_config = {
+       key: gdal.GetConfigOption(key) for key in new_config.keys()}
+   for key, val in new_config.items():
+       gdal.SetConfigOption(key, val)
+
+   # Perform GDAL operations in this context.
+   # ...
+   # Done.
+
+   # Restore previous configuration.
+   for key, val in prev_config.items():
+       gdal.SetConfigOption(key, val)
+
+Rasterio provides this with a single Python statement.
+
+.. code-block:: python
+
+   with rasterio.Env(CPL_DEBUG=True, GDAL_CACHEMAX=512):
+       # ...
+
```

Comment:
Maybe worth mentioning how environment variables interact with the GDAL environment. Specifically, there are lots of gdal examples in the wild that suggest setting gdal configs via env variables and we could explicitly discourage that practice.